target_decisioning_engine/notification_provider.py

```diff
@@ -0,0 +1,119 @@
+
+LOG_TAG = "{}.NotificationProvider".format(LOG_PREFIX)
+
+
+class NotificationProvider:
+    """NotificationProvider"""
+
+    def __init__(self, request, visitor, send_notification_func=noop, telemetry_enabled=True):
+        """
+        :param request: (delivery_api_client.Model.delivery_request.DeliveryRequest) request
+        :param visitor: (delivery_api_client.Model.visitor_id.VisitorId) VisitorId instance
+        :param send_notification_func: (callable) function used to send the notification
+        :param telemetry_enabled: (bool) is telemetry enabled
+        """
+        self.visitor = visitor
+        self.send_notification_func = send_notification_func
+        self.telemetry_enabled = telemetry_enabled
+        self.request = request
+        self.request_id = request.request_id
+        self.prev_event_keys = set()
+        self.notifications = []
+        self.telemetry_entries = []
+        self.logger = get_logger()
+
+    def add_notification(self, mbox, trace_func=noop):
+        """
+        :param mbox: (delivery_api_client.Model.mbox_response.MboxResponse) mbox
+        :param trace_func: (callable) trace function
+        """
+        display_tokens = []
+        for option in mbox.options:
+            event_token = option.event_token
+            event_key = "{}-{}".format(mbox.name, event_token)
+
+            if event_token and event_key not in self.prev_event_keys:
+                display_tokens.append(event_token)
+                self.prev_event_keys.add(event_key)
+
+        if not display_tokens:
+            return
+
+        notification_mbox = NotificationMbox(name=mbox.name)
+        notification = Notification(id=create_uuid(),
+                                    impression_id=create_uuid(),
+                                    timestamp=get_epoch_time_milliseconds(),
+                                    type=MetricType.DISPLAY,
+                                    mbox=notification_mbox,
+                                    tokens=display_tokens)
+        if callable(trace_func):
+            trace_func(notification)
+
+        self.notifications.append(notification)
+
+    def add_telemetry_entry(self, entry):
+        """
+        :param entry: (delivery_api_client.Model.telemetry_entry.TelemetryEntry) telemetry entry
+        """
+        if not self.telemetry_enabled:
+            return
+
+        entry.request_id = self.request_id
+        entry.timestamp = get_epoch_time_milliseconds()
+        entry.features = TelemetryFeatures(decisioning_method=DecisioningMethod.ON_DEVICE)
+        self.telemetry_entries.append(entry)
+
+    def send_notifications(self):
+        """Send notifications via the send_notification_func"""
+        self.logger.debug("{}.send_notifications - Notifications: {} \nTelemetry Entries: {}"
+                          .format(LOG_TAG, self.notifications, self.telemetry_entries))
+
+        if self.notifications or self.telemetry_entries:
```

Comment:
Let's short-circuit to reduce nesting
```suggestion
        if not self.notifications and not self.telemetry_entries:
            return
```